### PR TITLE
Replaced IncomingCallService & old BroadcastReceiver with 1 global broadcast receiver

### PR DESCRIPTION
**Pull request by afalls-twilio**
## Submission Checklist

 - [ ] Updated the `CHANGELOG.md` to reflect any **feature**, **bug fixes**, or **known issues** made in the source code
 - [x] Tested code changes and observed expected behavior in the example app
 - [x] Performed a visual inspection of the `Files changed` tab prior to submitting the pull request for review to ensure proper usage of the style guide

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Description

Refactored to have a single message pump for all voice call related intents.

## Breakdown

- Reduced codebase by ~200 lines
- Removed IncommingCallService
- Removed BroadcastReceiver
- Added a global Broadcast Receiver called VoiceNotificationReceiver
- Cleaned up notification helper to use Androidx.*Compat libs to remove version checks

## Validation

- Ran a series of tests locally

## Additional Notes
None

### CODE REVIEW

**Review by kbagchiGWC — APPROVED** [review]
Code changes look good. Thanks for the code clean up and the consolidation. A lot of code changes in this PR - please make sure to get a proper QA cycle from the feature branch before merging to main.

File: android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java [hunk 1]
```diff
@@ -229,24 +216,21 @@ public void voice_getVersion(Promise promise) {
...
-
-          if (fcmToken == null) {
-            Log.d(TAG, "FCM token is \"null\".");
-            promise.reject("FCM token is \"null\".");
-          } else {
-            promise.resolve(fcmToken);
-          }
+      .addOnCompleteListener(task -> {
+        if (!task.isSuccessful()) {
+          Log.w(TAG, "Fetching FCM registration token failed", task.getException());
```
Comment (mhuynh5757) [inline android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java:221]: Can we use the `log` helper function here?

File: android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java [hunk 2]
```diff
@@ -445,70 +429,64 @@ public void call_getStats(String uuid,  Promise promise) {
...
-
-          // Log and toast
-          if (BuildConfig.DEBUG) {
-            Log.d(TAG, "Registering with FCM with token " + fcmToken);
-          }
-          RegistrationListener registrationListener = createRegistrationListener(promise);
-          Voice.register(token, Voice.RegistrationChannel.FCM, fcmToken, registrationListener);
+      .addOnCompleteListener(task -> {
+        if (!task.isSuccessful()) {
+          Log.w(TAG, "Fetching FCM registration token failed", task.getException());
```
Comment (mhuynh5757) [inline android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java:434]: Same as above, can we use the `log` helper function?

File: android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java [hunk 3]
```diff
@@ -445,70 +429,64 @@ public void call_getStats(String uuid,  Promise promise) {
...
-
-          // Log and toast
-          if (BuildConfig.DEBUG) {
-            Log.d(TAG, "Registering with FCM with token " + fcmToken);
-          }
-          UnregistrationListener unregistrationListener = createUnregistrationListener(promise);
-          Voice.unregister(token, Voice.RegistrationChannel.FCM, fcmToken, unregistrationListener);
+      .addOnCompleteListener(task -> {
+        if (!task.isSuccessful()) {
+          Log.w(TAG, "Fetching FCM registration token failed", task.getException());
```
Comment (mhuynh5757) [inline android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java:462]: Same as above, could we use the `log` helper function?

File: android/src/main/java/com/twiliovoicereactnative/VoiceNotificationReceiver.java [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+      case ACTION_ACCEPT_CALL:
+        handleAccept(
+          context,
+          intent.getParcelableExtra(Constants.INCOMING_CALL_INVITE),
+          intent.getStringExtra(Constants.UUID));
+        break;
+      case ACTION_REJECT_CALL:
+        handleReject(
+          context,
+          Objects.requireNonNull(intent.getParcelableExtra(Constants.INCOMING_CALL_INVITE)),
```
Comment (mhuynh5757) [inline android/src/main/java/com/twiliovoicereactnative/VoiceNotificationReceiver.java:67]: Should the other cases also use `Objects.requireNonNull`?

File: android/src/main/java/com/twiliovoicereactnative/VoiceNotificationReceiver.java [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+      acceptOptions,
+      new CallListenerProxy(uuid, context));
+    Storage.callMap.put(uuid, call);
+    Storage.callInviteMap.remove(uuid);
+
+    // handle if event spawned from JS
+    if (Storage.callAcceptedPromiseMap.containsKey(uuid)) {
+      Promise promise = Objects.requireNonNull(Storage.callAcceptedPromiseMap.remove(uuid));
+      promise.resolve(serializeCall(uuid, call));
+    }
```
Comment (mhuynh5757) [inline android/src/main/java/com/twiliovoicereactnative/VoiceNotificationReceiver.java:163]: Very cool, I like that we have a single code path now.

File: android/src/main/java/com/twiliovoicereactnative/VoiceNotificationReceiver.java (hunk 4 above)
Comment (afalls-twilio) [reply android/src/main/java/com/twiliovoicereactnative/VoiceNotificationReceiver.java:67]: Done

File: android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java [hunk 6]
```diff
@@ -431,7 +432,7 @@ public void voice_register(String token, Promise promise) {
     FirebaseMessaging.getInstance().getToken()
       .addOnCompleteListener(task -> {
         if (!task.isSuccessful()) {
-          Log.w(TAG, "Fetching FCM registration token failed", task.getException());
+          warning(TAG, "Fetching FCM registration token failed", task.getException());
```
[the remark this answers is not in the document]
Comment (afalls-twilio) [reply android/src/main/java/com/twiliovoicereactnative/TwilioVoiceReactNativeModule.java:435]: good catch!

**Pull request merged**